lib/transaction_files.c

```diff
@@ -652,124 +653,308 @@ collect_files(struct xbps_handle *xhp, xbps_dictionary_t d,
-		    "%s: failed to read binary package `%s': %s",
-		    pkgver, bpkg, strerror(rv));
-		goto out;
+	for (unsigned int i = 0; i < xbps_array_count(alts); i++) {
+		const char *alt;
+		int r;
+		if (!xbps_array_get_cstring_nocopy(alts, i, &alt))
+			return -EINVAL;
+		r = xbps_alternative_link(alt, linkpath, sizeof(linkpath), NULL, 0);
+		if (r < 0) {
+			return r;
+		}
+		xbps_dbg_printf(xhp, ">>>> provider=%s group=%s linkpath=%s\n", pkgname, group, linkpath);
+		r = collect_file(xhp, linkpath, 0, changer->pkgname, changer->pkgver, changer->idx, NULL,
+		    TYPE_LINK, false, false, false, true, NULL);
+		if (r == EEXIST) {
+			error = true;
+			continue;
+		} else if (r != 0) {
+			return -r;
+		}
 	}
+	if (error)
+		return -EEXIST;
+	return 0;
+}
 
-	for (uint8_t i = 0; i < 4; i++) {
-		const char *entry_pname;
-		int ar_rv = archive_read_next_header(ar, &entry);
-		if (ar_rv == ARCHIVE_EOF || ar_rv == ARCHIVE_FATAL)
-			break;
-		else if (ar_rv == ARCHIVE_RETRY)
+static int
+register_alternative_groups(struct xbps_handle *xhp,
+		xbps_dictionary_t alternatives,
+		const struct alternative_changer *changer,
+		xbps_dictionary_t pkg_alternatives
+		)
+{
+	xbps_array_t allkeys = xbps_dictionary_all_keys(pkg_alternatives);
+	for (unsigned int i = 0; i < xbps_array_count(allkeys); i++) {
+		xbps_object_t keysym = xbps_array_get(allkeys, i);
+		const char *group = xbps_dictionary_keysym_cstring_nocopy(keysym);
+		const char *first;
+		xbps_array_t providers = xbps_dictionary_get(alternatives, group);
+		bool found = false;
+		int r;
+
+		if (!providers) {
+			/*
+			 * There are no previous providers for the alternative
+			 * group, this package becomes the provider.
+			 */
+			providers = xbps_array_create();
+			xbps_array_add_cstring(providers, changer->pkgname);
+			if (!xbps_dictionary_set(alternatives, group, providers))
+				return errno ? -errno : -ENOMEM;
+			xbps_object_release(providers);
+			r = register_alternative_links(xhp, changer, changer->pkgname, group);
+			if (r < 0)
+				return r;
 			continue;
+		}
 
-		entry_pname = archive_entry_pathname(entry);
-		if ((strcmp("./files.plist", entry_pname)) == 0) {
-			filesd = xbps_archive_get_dictionary(ar, entry);
-			if (filesd == NULL) {
-				rv = EINVAL;
-				goto out;
+		/*
+		 * The alternative group already registers, add this package
+		 * as provider if it is not already part of it.
+		 */
+		for (unsigned int j = 0; j < xbps_array_count(providers); j++) {
+			const char *cur;
+			xbps_array_get_cstring_nocopy(providers, i, &cur);
+			if (strcmp(cur, changer->pkgname) == 0) {
+				found = true;
+				break;
 			}
-			rv = collect_files(xhp, filesd, pkgname, pkgver, idx,
-			    update, false, false, false);
-			xbps_object_release(filesd);
-			goto out;
 		}
-		archive_read_data_skip(ar);
+		if (!found)
+			xbps_array_add_cstring(providers, changer->pkgname);
+		if (!xbps_array_get_cstring_nocopy(providers, 0, &first)) {
+			/* XXX: does this need to be handled? */
```

Comment:
This means
- xbps_array_add_cstring failed, or
- first elem of array is not a string, i.e. file is corrupted
thus should return error.
On the other hand, `*_add`, `*_set` calls are rarely checked for success in xbps.